### PR TITLE
Restrict submissions of new legacy add-ons

### LINKED ISSUE #4234: Restrict legacy add-on submissions
Main bug for the actual restriction of new legacy add-ons submissions as described in https://blog.mozilla.org/addons/2016/11/23/add-ons-in-2017/

The idea is to prevent the submission of new non-webextensions extensions in devhub. The following caveats apply:
- The restriction should be behind a waffle, not active yet.
- Only *new* submissions. Updates should still be possible for now.
- Only add-ons with the *extension* type, that are not webextensions.
- Only add-ons targeting only Firefox, Firefox for Android, or both. This leaves a loophole for extensions supporting multiple applications like Thunderbird or Seamonkey, but we can tackle that separately.
- Both listed or unlisted should be affected
- Only add-ons supporting version 53 or above (more precisely, we would only accept legacy Firefox extensions with strict compatibility and a `maxVersion` lower than 53.0)
- It should still be possible for an admin to click on “Override failed validation” in order to continue with the submission process.

### PR DESCRIPTION
**Pull request by diox**
Behind a waffle and several conditions.

Fix #4234

### CODE REVIEW

File: src/olympia/devhub/tasks.py [hunk 1]
```diff
@@ -267,6 +271,52 @@ def handle_file_validation_result(results, file_id, annotate=True):
...
+def annotate_new_legacy_addon_restrictions(results):
+    """
+    Annotate validation results to restrict uploads of new legacy
+    (non-webextension) add-ons if specific conditions are met.
+    """
+    metadata = results.get('metadata', {})
+    target_apps = metadata.get('applications', {})
+    target_firefox_version = (
+        target_apps.get('firefox', {}).get('max', None) or
+        target_apps.get('android', {}).get('max', None)
```
Comment (wagnerand) [inline src/olympia/devhub/tasks.py]: I think I remember we want to enable the restriction for Firefox Desktop only. Summoning @jvillalobos to confirm (or deny)...

Comment (diox) [inline src/olympia/devhub/tasks.py]: The [PRD](https://docs.google.com/document/d/1BdiJSYD4aOMClasGxx5wVShHEk8VbxefE3Gf_6jpgkQ/edit#) says *"that support only Firefox, Firefox for Android, or both"*

Comment (wagnerand) [inline src/olympia/devhub/tasks.py]: Oh, nevermind then.

File: src/olympia/devhub/tasks.py [hunk 2]
```diff
@@ -267,6 +271,52 @@ def handle_file_validation_result(results, file_id, annotate=True):
...
 
+def annotate_new_legacy_addon_restrictions(results):
+    """
+    Annotate validation results to restrict uploads of new legacy
+    (non-webextension) add-ons if specific conditions are met.
+    """
+    metadata = results.get('metadata', {})
+    target_apps = metadata.get('applications', {})
+    target_firefox_version = (
+        target_apps.get('firefox', {}).get('max', None) or
```
Comment (wagnerand) [inline src/olympia/devhub/tasks.py]: Actually, a new submission having maxversions like Firefox Android 54 and Desktop 51 would pass this. Admittedly, a little of an edge-case...

File: src/olympia/devhub/tasks.py [hunk 3]
```diff
@@ -290,9 +290,12 @@ def annotate_new_legacy_addon_restrictions(results):
         set(target_apps.keys())
     )
     is_targeting_firefox_lower_than_53_only = (
-        target_firefox_version and
         metadata.get('strict_compatibility') is True and
-        version_int(target_firefox_version) < 53000000200100
+        # version_int('') is actually 200100. If strict compatibility is true,
+        # the validator should have complained about the non-existant max
+        # version, but it doesn't hurt to check that the value is sane anyway.
+        max_target_firefox_version > 200100 and
```
Comment (wagnerand) [inline src/olympia/devhub/tasks.py]: 🌟

**Pull request merged**